test/integration/sessions_test.rb

```diff
@@ -1,54 +1,171 @@
+        and_i_see_an_error_about_invalid_cookie_consent
+
+        when_i_consent_to_cookies
+        assert_this_redirects_me { and_i_click_continue }
+        and_my_consents_have_been_saved
+      end
+
+      should "Enforce that feedback consent is explicitly given or rejected" do
+        when_i_return_from_digital_identity
+        then_i_see_the_first_time_form
+
+        when_i_consent_to_cookies
+        and_i_click_continue
+        then_i_see_the_first_time_form
+        and_i_see_an_error_about_invalid_feedback_consent
+
+        when_i_consent_to_feedback
+        assert_this_redirects_me { and_i_click_continue }
+        and_my_consents_have_been_saved
+      end
+    end
+  end
+
+  context "Given a returning user" do
+    should "Log the user in and send them to the account dashboard" do
+      given_a_successful_login_attempt(cookie_consent: true, feedback_consent: false)
+      assert_this_redirects_me { when_i_return_from_digital_identity }
+      and_i_see_the_dashboard
+    end
+
+    context "With a redirect path" do
+      setup do
+        @redirect_path = "/email/subscriptions/account/confirm?frequency=immediately&return_to_url=true&topic_id=some-page-with-notifications"
+        given_a_successful_login_attempt(cookie_consent: true, feedback_consent: false)
+      end
+
+      should "Log the user in and send them to the redirect path" do
+        assert_this_redirects_me { when_i_return_from_digital_identity }
+      end
+    end
+  end
 
-      stub_account_api_set_attributes(
+  def given_a_successful_login_attempt(cookie_consent:, feedback_consent:)
+    @govuk_account_session = "new-session-id"
+
+    stub_account_api_validates_auth_response(
+      govuk_account_session: @govuk_account_session,
+      redirect_path: @redirect_path,
+      cookie_consent: cookie_consent,
+      feedback_consent: feedback_consent,
+    )
+
+    if cookie_consent.nil? || feedback_consent.nil?
+      @stub_set_consents = stub_account_api_set_attributes(
         attributes: {
           cookie_consent: true,
           feedback_consent: true,
         },
         govuk_account_session: @govuk_account_session,
         new_govuk_account_session: @govuk_account_session,
       )
-    end
 
-    should "send the user back to the redirect path" do
+      # needed because we can't make capybara handle our special
+      # response header (ideally we'd have some code run between
+      # visiting the callback endpoint and making the request to the
+      # form, to pluck out the response header and fill in the request
+      # header)
       mock_logged_in_session "!#{@govuk_account_session}"
+    end
 
-      visit new_govuk_session_callback_path(code: "code", state: "state")
-
-      assert page.has_content? "use cookies to collect anonymised analytics"
+    @stub_user_info = stub_account_api_user_info
+    stub_email_alert_api_authenticate_subscriber_by_govuk_account("!#{@govuk_account_session}", "subscriber-id", "email@example.com")
```

Comment:
Right so the proposal is we move the need for three three bits, integrating the data we've stubbed here into GET /api/user.

Makes sense to me.

Reply:
Yeah, if we return `services: { notifications: "yes" }` or `services: { notifications: "no" }` from `GET /api/user` then the controller here can just read that.